Packet++/src/SomeIpLayer.cpp

```diff
@@ -0,0 +1,349 @@
+#define LOG_MODULE PacketLogModuleSomeIpLayer
+
+#include "SomeIpLayer.h"
+#include "EndianPortable.h"
+#include "Packet.h"
+#include "PayloadLayer.h"
+#include <algorithm>
+#include <sstream>
+#include <string.h>
+
+namespace pcpp
+{
+std::unordered_set<uint16_t> SomeIpLayer::mSomeIpPorts{};
+
+void SomeIpLayer::optionallyCopyPayloadData(const uint8_t *const data, size_t dataLen, bool dontCopy)
+{
+	const size_t headerLen = sizeof(someiphdr);
+
+	if (dontCopy)
+	{
+		m_DataLen = headerLen;
+	}
+	else
+	{
+		m_DataLen = headerLen + dataLen;
+	}
+
+	m_Data = new uint8_t[m_DataLen];
+	m_Protocol = SomeIp;
+	memset(m_Data, 0, headerLen);
+
+	if (dontCopy)
+	{
+		m_PayloadPtr = (uint8_t *)data;
+	}
+	else
+	{
+		memcpy(m_Data + headerLen, data, dataLen);
+		m_PayloadPtr = nullptr;
+	}
+
+	setPayloadLength(dataLen);
+}
+
+SomeIpLayer::SomeIpLayer(uint16_t serviceID, uint16_t methodID, uint16_t clientID, uint16_t sessionID,
+						 uint8_t interfaceVersion, MsgType type, uint8_t returnCode, const uint8_t *const data,
+						 size_t dataLen, bool dontCopy)
+{
+	optionallyCopyPayloadData(data, dataLen, dontCopy);
+	setServiceID(serviceID);
+	setMethodID(methodID);
+	setClientID(clientID);
+	setSessionID(sessionID);
+	setInterfaceVersion(interfaceVersion);
+	setMessageType(type);
+	setReturnCode(returnCode);
+}
+
+SomeIpLayer::SomeIpLayer(const SomeIpLayer &other) : Layer(other)
+{
+	copyLayerData(other);
+}
+
+void SomeIpLayer::copyLayerData(const SomeIpLayer &other)
+{
+	m_PayloadPtr = other.m_PayloadPtr;
+}
+
+SomeIpLayer &SomeIpLayer::operator=(const SomeIpLayer &other)
+{
+	Layer::operator=(other);
+
+	copyLayerData(other);
+
+	return *this;
+}
+
+uint32_t SomeIpLayer::getLengthField() const
+{
+	return be32toh(getSomeIpHeader()->length);
+}
+
+void SomeIpLayer::setPayloadLength(uint32_t payloadLength)
+{
+	someiphdr *hdr = getSomeIpHeader();
+	hdr->length = htobe32(sizeof(someiphdr) - sizeof(hdr->serviceID) - sizeof(hdr->methodID) - sizeof(hdr->length) +
+						  payloadLength);
+}
+
+uint32_t SomeIpLayer::getHeaderLengthWithPayload() const
+{
+	/* Add outer header of someip to length (messageId and length field itself) */
+	return sizeof(uint32_t) * 2 + getLengthField();
+}
+
+size_t SomeIpLayer::getHeaderLen() const
+{
+	if (subProtocolFollows())
+	{
+		/* If there is a protocol coming after the normal SomeIp header, return just the header length so PcapPlusPlus
+		handles the oncoming protocols correctly */
+		return sizeof(someiphdr);
+	}
+	else
+	{
+		/* Else, return the length of the header + payload, since it is a normal SomeIp PDU */
+		return getHeaderLengthWithPayload();
+	}

```

Comment:
@seladb: we do it this way because there can be more SOME/IP (SOME/IP-TP) frames in one packet, if getHeaderLen() always returns sizeof(someiphdr) and PayloadLayer whould contain the rest, this whould not be possible because the PayloadLayer is always the last layer. Currently subProtocolFollows() returns always false but this will be extended with SOME/IP-TP/SD. What do you think about this solution, could we let it that way?

Reply:
The way PcapPlusPlus packet parser works is that each layer is responsible for parsing the next layer, via implementing the `parseNextLayer()` method.
In this PR SOME/IP-TP and SOME/IP-SD are still missing, hence we can do the following:
- Check if the next layer is SOME/IP-TP or SOME/IP-SD
  - If yes - create a `PayloadLayer` (since we don't know how to parse these layers yet)
  - If no - check `SomeIpLayer::isDataValid(...)`:
     - If yes - create a `SomeIpLayer`
     - If no - create a `PayloadLayer` 

With this design we can keep `getHeaderLen()` very simple and can always return `sizeof(someiphdr)`, and `SomeIpLayer::parseNextLayer()` does the more heavy lifting.

Please let me know what you think.

Reply:
i updated the Branch with adding the PayloadLayer for SD/TP frames

for the getHeaderLen() method i am not sure how you how you mean this because in a SOME/IP frame there is always the header + payload and when there are two SOME/IP frames in one packet than there is header + payload + header + payload (i added a test for such a frame and updated the pcap file). So in my opinion this method can not be reduced to only sizeof(someiphdr) because it must also contain the length of the payload when it is a SOME/IP frame. Can you explain it again please?

Reply:
Oh ok, I missed that SOME/IP frames can have a payload, I'm sorry about that. In that case, your code is correct